### PR TITLE
Allows the counting of number of records in large file

### PR DESCRIPTION
**Pull request by qchaupds**
* Modify logic to count the number of records in a large file in TableReader.java
Refs:

https://github.com/NASA-PDS/validate/issues#326 File-size check fails for large data files
https://github.com/NASA-PDS/validate/issues#327 validate fails to process large data file

The testing will be done via validate pull request which uses the service of TableReader.java
This pull request should be merged before validate since there is a dependency.

### CODE REVIEW

**Comment by qchaupds** [comment]
Please hold off on merging.   For some strange reason, validate regression test is hanging.  Trying to figure out who is the culprit.

File: src/main/java/gov/nasa/pds/objectAccess/TableReader.java [hunk 1]
```diff
@@ -311,6 +310,42 @@ public int getCurrentRow() {
 	public ByteWiseFileAccessor getAccessor() {
 	  return this.accessor;
 	}
+
+    private long countCarriageReturn(URL dataFile) throws Exception {
```
Comment (jordanpadams) [inline src/main/java/gov/nasa/pds/objectAccess/TableReader.java:314]: @qchaupds just curious, when we do content validation, aren't we using this ByteBuffer then as well? In which case, doesn't this mean we are now going through and reading in the entire file multiple times? Could we combine this with content validation? It seems like if we have to open up the actual product to get some information (e.g. record size), then that could be considered in the realm of content validation.

File: src/main/java/gov/nasa/pds/objectAccess/TableReader.java [hunk 2]
```diff
@@ -325,14 +360,11 @@ public long getRecordSize(URL dataFile, Object table) throws Exception {
...
-			bufferedReader = new BufferedReader(new InputStreamReader(is, "US-ASCII"));
-			int numLines = 0;
-			while (bufferedReader.readLine()!=null) {
-			   ++numLines;
-			}
-			this.recordSize = numLines;
+            // Removed old usage of BufferedReader to count lines which is problematic when file is large.
+            // Use new function countCarriageReturn() to count number of lines of any file sizes.
+
+            this.recordSize = this.countCarriageReturn(dataFile);
```
Comment (jordanpadams) [inline src/main/java/gov/nasa/pds/objectAccess/TableReader.java:366]: @qchaupds per our convo today and [this ticket](https://github.com/NASA-PDS/validate/issues/325), I don't think this method will work in the event the CRLF is embedded within a record. we should consider checking for record count using opencsv when we are already reading in the data to do content validation.

**Review by jordanpadams — CHANGES_REQUESTED** [review]

File: src/main/java/gov/nasa/pds/objectAccess/TableReader.java (hunk 2 above)
Comment (jordanpadams) [reply src/main/java/gov/nasa/pds/objectAccess/TableReader.java]: from @qchaupds The code change in TableReader.java in pds4-jparser now uses the record length to calculate the number of records.  No need to read the file.

File: src/main/java/gov/nasa/pds/objectAccess/TableReader.java [hunk 3]
```diff
@@ -311,42 +315,128 @@ public int getCurrentRow() {
...
+    }
 	
 	/**
 	 * @return the size of record (i.e. number of lines)
 	 */
 	public long getRecordSize(URL dataFile, Object table) throws Exception {
 		adapter = AdapterFactory.INSTANCE.getTableAdapter(table);
 		InputStream is = Utility.openConnection(dataFile.openConnection());
+
+        LOGGER.debug("getRecordSize:adapter {}",adapter);
```
Comment (jordanpadams) [inline src/main/java/gov/nasa/pds/objectAccess/TableReader.java:407]: @qchaupds hate to nitpick here again, but can we fix the whitespace here and elsewhere in your code updates to lineup all the tabs correctly for better readability?

**Review by jordanpadams — CHANGES_REQUESTED** [review]

File: src/main/java/gov/nasa/pds/objectAccess/TableReader.java [hunk 4]
```diff
@@ -311,42 +315,128 @@ public int getCurrentRow() {
...
 	  return this.accessor;
 	}
+
+    private long parseBufferForLineCount(URL dataFile, byte [] bufferAsBytes) throws Exception {
+        // Given a byte array, read through as if reading through a smaller file and count the lines.
+        InputStream inputStream = new ByteArrayInputStream(bufferAsBytes);
+        InputStreamReader inputStreamReader = new InputStreamReader(inputStream);
+        long linesInBuffer = 0;
+        try (BufferedReader reader = new BufferedReader(inputStreamReader)) {
+            while (reader.readLine() != null) linesInBuffer++;
```
[the remark this answers is not in the document]
Comment (jordanpadams) [reply src/main/java/gov/nasa/pds/objectAccess/TableReader.java:325]: per email from @qchaupds 

per the middle of the line question, The buffer is a chunk of a file up to so many bytes.  That chunk is given to the parseBufferForLineCount().  The function returns the number of lines read via readLine().   Once that first chunk is done, the next part of the file is then fetched into another chunk.
 
The regression tests fortunately catches any mis-counting of the number of records.

per the `readline` question, the documentation says it handles both cases well.
 
https://docs.oracle.com/javase/8/docs/api/java/io/BufferedReader.html#readLine--
 
public String readLine()
                throws IOException
Reads a line of text. A line is considered to be terminated by any one of a line feed ('\n'), a carriage return ('\r'), or a carriage return followed immediately by a linefeed.

**Review by jordanpadams — APPROVED** [review]

**Pull request merged**